apps/prairielearn/src/sync/fromDisk/assessments.js

```diff
@@ -291,14 +292,42 @@ function parseSharedQuestionReference(qid) {
   };
 }
 
+/**
+ * Determines if a course instance is accessible. A course instance is considered
+ * to be accessible if any access rules either have no end date or have an end date
+ * in the future.
+ *
+ * @param {import('../course-db').CourseInstanceData} courseInstanceData
+ */
+function isCourseInstanceAccessible(courseInstanceData) {
+  const courseInstance = courseInstanceData.courseInstance.data;
+
+  // If the course instance data is not available, treat it as though it's
+  // not accessible.
+  if (!courseInstance) return false;
+
+  // If there are no access rules, the course instance is not accessible.
+  if (!courseInstance.allowAccess?.length) return false;
+
+  return courseInstance.allowAccess.some((accessRule) => {
+    if (!accessRule.endDate) return true;
+    return isFuture(parseISO(accessRule.endDate));
```

Comment:
This is likely pedantic, but `endDate` is not ISO. I doubt this would be a problem (since an end-date within 24 hours of the current date is unlikely to affect this result significantly), but it might be good to either take the CI's timezone into account when parsing, compare it against the current time in the CI's timezone, explicitly subtract one day from the result, or add a comment that this is known and being ignored.

Reply:
That's fair. I copied the `parseISO` and `isFuture` from existing syncing code (though of course it's code that I wrote, so no surprise that I missed this subtlety before):

https://github.com/PrairieLearn/PrairieLearn/blob/34c498b4639bb1acef013da485c9e19cf0a274a4/apps/prairielearn/src/sync/course-db.ts#L946-L980

I'm inclined to just add a comment. The likelihood that anyone added a valid legacy scheduler UUID in the last day, let alone the last year, is extremely low.

Reply:
I added comments in 8746a2e. I also made a few other changes that felt appropriate:

- Parsing is now done with `new Date` to reflect that we're not specifically parsing an ISO string.
- I duplicated the validation done by the `input_date` sproc higher up in the stack to ensure that we're always showing the best possible error messages to users.

Reply:
> I duplicated the validation done by the input_date sproc higher up in the stack to ensure that we're always showing the best possible error messages to users.

I noticed that you have a regexp search instead of a full-string match, though it seems this is also the case in input_date. This means that if the string includes text that is not part of the date, this text is ignored. In `parseISO` this would throw a sync error. That said, a full-string match would also reject other suffixes (e.g., `Z` or `+00:00`) that are accepted in `parseISO`. So:

String | Before | After
--- | --- | ---
`2024-03-15T14:03:21` | Accepted | Accepted
`2024-03-15T14:03:21Z` | Accepted | Accepted
`2024-03-15T14:03:21Z WITH MORE TEXT` | Rejected | Accepted
`2024-03-15T14:03:21+00:00` | Accepted | Accepted
`2024-03-15T14:03:21 2024-03-16T14:03:21` | Rejected | Accepted

Note that some of these cases will be accepted by the new function that checks for validity, but will cause problems in the `new Date` case.

I'm fine if you decide this is not worth spending time on, I just wanted to point this out.

Reply:
That's fair. I added the regexp to ensure that we'd catch something like `2024-04` early, as `parseISO` and `new Date` will happily accept that. I think I'll keep the regexp, but also move back to `parseISO` the entire string. Would that cover all our bases?

Reply:
I think that should work.

Reply:
Done in 31edfbd.